Use __index__() method instead of checking explicit list of types

namannimmo10 opened the pull request:
… types

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #16779 

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

namannimmo10 commented:
@oscarbenjamin Should I add `np.int64` to `SYMPY_INTS`?

oscarbenjamin commented on sympy/core/compatibility.py:
```diff
@@ -396,15 +396,14 @@ def as_int(n, strict=True):
         return result
-    except TypeError:
+    except AttributeError:
         raise ValueError('%s is not an integer' % (n,))
```
We shouldn't catch `AttributeError`.

Use the `operator.index()` function instead which will raise `TypeError` instead. We don't need to use `int` if `index` returns a value. There should be no need to check the value returned by index either:
https://www.python.org/dev/peps/pep-0357/

namannimmo10 commented:
I've tried this diff:
``` >@@ -397,13 +397,10 @@ def as_int(n, strict=True):
     ValueError: ... is not an integer
     """
     try:
-        if strict and n.__index__() != n:
+        if strict and not hasattr(n, '__index__'):
             raise TypeError
-        result = int(n)
-        if result != n:
-            raise TypeError
-        return result
-    except AttributeError:
+        return n.__index__()
+    except TypeError:
         raise ValueError('%s is not an integer' % (n,))
```

And it works: 
```
>>> import numpy as np
>>> from sympy.core.compatibility import as_int
>>> as_int(np.int64(123))
123
>>> type(_)
<class 'int'>
```

namannimmo10 commented:
Ping @oscarbenjamin

oscarbenjamin commented:
What happens if `strict` is False? It's easier for me to understand what you are proposing if you push it to the PR. You can always squash the commits (now or later).

namannimmo10 commented:
@oscarbenjamin, Why does it show this exception, given the fact that I have numpy installed on my system. 
>    ModuleNotFoundError: No module named 'numpy'

oscarbenjamin commented:
> @oscarbenjamin @smichr Why does it show this exception, given the fact that I have numpy installed on my system.
> 
> > ModuleNotFoundError: No module named 'numpy'

That error is on Travis not on your system. It doesn't matter whether or not numpy is installed on your computer.

asmeurer commented on sympy/core/tests/test_compatibility.py:
```diff
@@ -4,6 +4,7 @@
 from sympy.testing.pytest import raises
 
 from sympy.abc import x
+import numpy as np
```
You need to allow this test to fail if numpy is not installed. Take a look at how it is done in sympy/utilities/tests/test_lambdify.py

namannimmo10 replied:
Alright, I'll look into it. Thanks.

asmeurer commented:
We shouldn't add np.int64 to the sympy ints because that would require importing numpy, and we don't want to make numpy a required dependency of sympy, nor do we want to import it when we import sympy.

asmeurer commented:
The current code seems correct based on https://docs.python.org/3/reference/datamodel.html#object.__index__: 

> Presence of this method indicates that the numeric object is an integer type. Must return an integer.

> If `__int__()`, `__float__()` and `__complex__()` are not defined then corresponding built-in functions `int()`, `float()` and `complex()` fall back to `__index__()`.

oscarbenjamin commented:
> The current code seems correct based on https://docs.python.org/3/reference/datamodel.html#object.__index__:

The officially supported way to get an `int` with the integer value of an object that must be an integer type and not e.g. an integer-valued float is to use index:
https://www.python.org/dev/peps/pep-0357/

Then you don't need an explicit list of types and you don't have to check anything else.

The mistake in this PR right now is that it checks for the presence of `__index__` but then calls `int`. It should just be
```
from operator import index
...
    if strict:
        try:
            return index(n)
        except TypeError:
            raise ValueError(...)
```

namannimmo10 commented:
> That error is on Travis not on your system. It doesn't matter whether or not numpy is installed on your computer.

But in [this](https://github.com/sympy/sympy/pull/18192#issuecomment-576734546) PR, I just updated my pip version, closed the issue, reopened it and then the build was successful after that. 
...........................................................................
> The mistake in this PR right now is that it checks for the presence of `__index__` but then calls `int`. 

That's bc I need to deal with the condition when `strict` is False....  
Consider this example:
```
>>> big = 1e23
>>> type(big) is float
True
>>> big == int(big)
True
>>> as_int(big)
Traceback (most recent call last):
 ...
 ValueError: ... is not an integer
 >>> as_int(big, strict=False)
 99999999999999991611392
```
...................................................................................

> It should just be
> 
> ```
> from operator import index
> ...
>     if strict:
>         try:
>             return index(n)
>         except TypeError:
>             raise ValueError(...)
> ```
What about the case when `strict` is False?? I guess you need to call `int` in the `else` block. The latest commit I've added passes all the tests.  Please correct me if I'm wrong, @oscarbenjamin @asmeurer.

namannimmo10 commented:
@asmeurer @oscarbenjamin, Please let me know what needs to be done.

asmeurer commented on sympy/core/tests/test_compatibility.py:
```diff
@@ -77,8 +79,5 @@ def test_ordered():
-    assert round(S(12.3)).is_Integer
-    assert round(S(12.3), 0).is_Float
+def test_issue_16779():
+    assert as_int(numpy.int64(123)) == 123
```
This  needs to skip the test if `numpy` is None.

asmeurer commented:
> What about the case when strict is False?? I guess we need to call int in the else block. The latest commit I've added passes all the tests. Please correct me if I'm wrong, @oscarbenjamin @asmeurer.

I think it should be the same in the `else` block as it is now. So if `strict` is `True` it would call `index` and if it is `False` it would call `int`.  We can also document that that is the behavior in the docstring.

namannimmo10 commented:
Ready for review :)

namannimmo10 commented:
It's ready now. Please merge this if looks good.

oscarbenjamin commented on sympy/core/compatibility.py:
```diff
@@ -396,16 +396,18 @@ def as_int(n, strict=True):
+        try:
+            if n != int(n):
+                raise TypeError
+            return int(n)
```
This calls `int(n)` twice. It would be better to call it only once.

oscarbenjamin commented on sympy/core/compatibility.py:
```diff
@@ -396,16 +396,18 @@ def as_int(n, strict=True):
+                raise TypeError
+            return int(n)
+        except TypeError:
+            raise ValueError('%s is not an integer' % (n,))
```
Here we raise TypeError so that we can catch it and then raise ValueError. It would be better just to raise ValueError in the first place and then the outer try is not needed.

namannimmo10 commented:
Thanks for the review!

oscarbenjamin commented:
It looks like there are now many errors because the previous code was catching `TypeError` from `int(n)`. Probably it should be caught again but using a try that is only around `int(n)`.

namannimmo10 commented:
@asmeurer @oscarbenjamin ready to merge.

asmeurer commented on sympy/core/compatibility.py:
```diff
@@ -352,8 +352,8 @@ def as_int(n, strict=True):
     is raised if the input has a non-integral value. When ``strict`` is
-    False, non-integer input that compares equal to the integer value
-    will not raise an error.
+    False, it would call ``int`` in the ``else`` block and non-integer
```
```suggestion
    True this uses `__index__ <https://docs.python.org/3/reference/datamodel.html#object.__index__>`_
    and when it is False it uses `int`.
```

asmeurer replied:
You should probably build the Sphinx docs to make sure that formats correctly.

namannimmo10 replied:
> You should probably build the Sphinx docs to make sure that formats correctly.

Great tool.. thanks @asmeurer :)

asmeurer commented on sympy/core/compatibility.py:
```diff
@@ -352,8 +352,8 @@ def as_int(n, strict=True):
-    False, non-integer input that compares equal to the integer value
-    will not raise an error.
+    False, it would call ``int`` in the ``else`` block and non-integer
+    input that compares equal to the integer value will not raise an error.
```
```suggestion

```

namannimmo10 replied:
Thanks!

asmeurer commented:
I guess this needs some sort of test.

asmeurer commented:
Looking at this, after this change it isn't clear to me what the point of `as_int` is with strict=True (the default). Would it be better to just use `operator.index` in the codebase? The only difference in the current code is that it converts TypeError to ValueError, which I don't really understand the need for. Is there a good reason to keep `as_int(strict=True)`? 

I guess for strict=False the difference is it allows floats that are exact integers. It looks like strict=False is only used in a few places in the code (most notably in evalf.py).

namannimmo10 commented:
> I guess this needs some sort of test.

What do you mean by that? Testing the docstring?

oscarbenjamin commented:
I'm also unclear on what the point is of `as_int` with strict=False. In #16081 the behaviour was changed to strict=True by default and places that were depending on non-strictness had strict=False added. I wonder for each of the places where strict=False is used whether something else should be done instead though:
```
$ git grep as_int | grep strict
sympy/core/compatibility.py:def as_int(n, strict=True):
sympy/core/compatibility.py:    >>> as_int(big, strict=False)
sympy/core/evalf.py:                # use strict=False with as_int because we take
sympy/core/evalf.py:                        as_int(v, strict=False)
sympy/core/evalf.py:                            [as_int(i, strict=False) for i in v.as_real_imag()]
sympy/core/power.py:                        as_int(pow, strict=False)
sympy/functions/special/gamma_functions.py:        as_int(n, strict=False)
```
From those `gamma_functions` uses it to define an `intlike` function that could just as easily have its own logic. There is one use in power that could possibly use something different. The others are in evalf which could have its own function for doing this.

namannimmo10 commented:
I agree with both of you. 
> From those gamma_functions uses it to define an intlike function that could just as easily have its own logic. There is one use in power that could possibly use something different. The others are in evalf which could have its own function for doing this.

But doing something different for all of them will take a considerable amount of time. I'm not sure it's worth it.  Please let me know what do you think and what needs to be done right now.

namannimmo10 commented:
Ping @oscarbenjamin @asmeurer

namannimmo10 commented:
@smichr what's your opinion on this?

oscarbenjamin commented:
Can you squash the commits? I think it's fine to merge this and make other changes later.

Probably it would be better to have two distinct functions rather than a strict=True flag. @asmeurer suggested using index directly instead of `as_int`. I haven't looked through all uses to know whether that makes sense (index could be a cryptic name in some places) but having a direct alias for index with a reasonable name makes sense.

namannimmo10 committed:
Update as_int function and remove PY2 related test

namannimmo10 commented:
OK. I just squashed it into a single commit.

oscarbenjamin commented:
Thanks. Merging.

asmeurer commented:
For naming I would just `import operator` and use `operator.index`, so it is clear what is being used.
